### PR TITLE
feat(android): robust download watchdog, retry logic, and battery optimization prompt

### PR DESCRIPTION
**Pull request by dishit-wednesday**
## Summary

- **Stuck-download watchdog**: polls every 15s, triggers retry after 3 consecutive zero-progress checks (45s of zero bytes = stuck). Network loss pauses the watchdog so legitimate network pauses are never misidentified as stuck.
- **Retry with backoff**: restarts the download up to 3 times with exponential backoff (30s, 60s, 90s). After max retries, fires `DownloadError` to JS.
- **Battery optimization prompt**: checks if app is excluded from OEM battery optimization before each download. If not, shows a one-time dialog prompting the user to whitelist the app — fixes silent download throttling on Honor, Huawei, Xiaomi devices.
- **Refactor for testability**: extracted stuck detection into a pure `evaluateStuckProgress()` companion function with a sealed `StuckAction` return type. Exposed watchdog constants as `internal`.
- **15 new unit tests**: cover watchdog constants, stuck threshold, retry escalation, give-up logic, counter reset on progress, and edge cases.

## Files changed

| File | Change |
|---|---|
| `AndroidManifest.xml` | Add `REQUEST_IGNORE_BATTERY_OPTIMIZATIONS` permission |
| `DownloadForegroundService.kt` | Logging improvements, `onDestroy` lifecycle hook |
| `DownloadManagerModule.kt` | Watchdog, network callback, retry logic, battery optimization methods |
| `DownloadManagerModuleTest.kt` | 15 new test cases for watchdog and stuck detection logic |
| `backgroundDownloadService.ts` | `checkAndPromptBatteryOptimization()` called before every download |

## Test plan

- [ ] Run `npm run test:android` — all download unit tests pass
- [ ] Start a model download, background the app, lock screen — verify download continues
- [ ] On first download, battery optimization dialog appears on Android devices not already whitelisted
- [ ] Simulate network drop during download — verify watchdog does not trigger (network lost pauses it)
- [ ] Verify after 45s of zero progress with network connected, download retries automatically
- [ ] After 3 failed retries, verify `DownloadError` event surfaces to the UI

### CODE REVIEW

File: android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt [hunk 1]
```diff
@@ -250,12 +333,14 @@ class DownloadManagerModule(reactContext: ReactApplicationContext) :
...
         try {
             val id = downloadId.toLong()
+            android.util.Log.d("DownloadService", "Cancelling download - id: $id")
 
             // Get download info BEFORE removing from SharedPreferences
             val downloadInfo = getDownloadInfo(id)
 
             downloadManager.remove(id)
             removeDownload(id)
+            downloadBytesTracker.remove(id)
```
[the remark this answers is not in the document]
Comment (dishit-wednesday) [reply android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt]: Fixed — wrapped `downloadBytesTracker.remove(id)` in `handler.post { ... }` to ensure it runs on the main thread, consistent with all other accesses to this map.

File: android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt [hunk 2]
```diff
@@ -416,6 +505,230 @@ class DownloadManagerModule(reactContext: ReactApplicationContext) :
...
+
+                // Seed tracker with current retry count so the watchdog knows how many attempts remain
+                handler.post { downloadBytesTracker[newDownloadId] = BytesTrack(0L, 0, retryCount) }
+
+                android.util.Log.d("DownloadService", "Re-enqueued - new id: $newDownloadId (replaced: $downloadId, retry $retryCount/$MAX_RETRY_ATTEMPTS)")
+
+            } catch (e: Exception) {
+                android.util.Log.e("DownloadService", "Failed to restart stuck download $downloadId: ${e.message}")
+            }
+        }
```
[the remark this answers is not in the document]
Comment (dishit-wednesday) [reply android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt:761]: Fixed — removed `Thread.sleep()` from the executor. The re-enqueue logic is now split: file cleanup runs on the executor (IO), then `watchdogHandler.postDelayed()` handles the backoff and re-enqueue on the main handler thread. This frees the executor immediately and does not block other downloads.

File: android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt [hunk 3]
```diff
@@ -416,6 +516,247 @@ class DownloadManagerModule(reactContext: ReactApplicationContext) :
...
+                    android.util.Log.d("DownloadService", "Deleting partial file (${partialFile.length()}B): ${partialFile.name}")
+                    partialFile.delete()
+                }
+
+                // Step 2: backoff using postDelayed — does NOT block the executor thread
+                val backoffMs = retryCount * 30_000L
+                android.util.Log.d("DownloadService", "Backoff ${backoffMs / 1000}s before re-enqueue (retry $retryCount/$MAX_RETRY_ATTEMPTS)")
+
+                watchdogHandler.postDelayed({
+                    try {
```
[the remark this answers is not in the document]
Comment (dishit-wednesday) [reply android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt]: Fixed — added a guard at the top of the postDelayed callback: if getDownloadInfo(downloadId) returns null, the download was cancelled during the backoff and we return early, preventing resurrection. Additionally, downloadManager.enqueue() and persistDownload() are now wrapped inside executor.execute within the postDelayed callback so all disk IO (file deletion, enqueue, persist) runs off the main thread.

File: android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt [hunk 4]
```diff
@@ -25,6 +34,10 @@ class DownloadManagerModule(reactContext: ReactApplicationContext) :
         const val PREFS_NAME = "OffgridMobileDownloads"
         const val DOWNLOADS_KEY = "active_downloads"
         private const val POLL_INTERVAL_MS = 500L
+        internal const val WATCHDOG_INTERVAL_MS = 15_000L  // Check every 15 seconds
+        internal const val STUCK_THRESHOLD = 3             // 3 × 15s = 45 seconds of zero progress after first byte
```
[the remark this answers is not in the document]
Comment (dishit-wednesday) [reply android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt:38]: Acknowledged the trade-off. The 45 s window applies only to mid-download freezes (after at least one byte has been received), so short stalls are naturally buffered by the STUCK_THRESHOLD=3 counter — two consecutive no-progress polls are ignored, the download is only restarted on the third. The startup path is handled separately with STARTUP_TIMEOUT_POLLS=8 (2 minutes) before an error is surfaced. The retry also deletes only the partial temp file (which the DownloadManager already placed in external cache), not user data, so the bandwidth cost is bounded by whatever was downloaded in that 45 s window — typically a few MB for a multi-GB model. We think this is the right balance given the OEM battery-saver environments these downloads run in, but have noted this as a potential tuning point if field data shows too many false-positive retries.

File: android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt [hunk 5]
```diff
@@ -142,10 +199,46 @@ class DownloadManagerModule(reactContext: ReactApplicationContext) :
...
+        }
+    }
+
     override fun getName(): String = NAME
 
     override fun onCatalystInstanceDestroy() {
         super.onCatalystInstanceDestroy()
+        isPolling = false
+        handler.removeCallbacks(pollRunnable)
+        watchdogHandler.removeCallbacks(watchdogRunnable)
```
[the remark this answers is not in the document]
Comment (dishit-wednesday) [reply android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt]: Fixed — changed to watchdogHandler.removeCallbacksAndMessages(null). This clears both the periodic watchdog runnable and any pending retry postDelayed callbacks scheduled by handleStuckDownload, ensuring no code from this module runs after the Catalyst instance is destroyed.

File: android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt [hunk 6]
```diff
@@ -416,6 +516,247 @@ class DownloadManagerModule(reactContext: ReactApplicationContext) :
...
+                // Step 2: backoff using postDelayed — does NOT block the executor thread
+                val backoffMs = retryCount * 30_000L
+                android.util.Log.d("DownloadService", "Backoff ${backoffMs / 1000}s before re-enqueue (retry $retryCount/$MAX_RETRY_ATTEMPTS)")
+
+                watchdogHandler.postDelayed({
+                    try {
+                        val request = DownloadManager.Request(Uri.parse(resolvedUrl))
+                            .setTitle(title)
+                            .setDescription("Downloading model...")
+                            .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
```
[the remark this answers is not in the document]
Comment (dishit-wednesday) [reply android/app/src/main/java/ai/offgridmobile/download/DownloadManagerModule.kt]: Partially addressed — the retry now reads description from the persisted downloadInfo via optString("description", "Downloading model..."). Full hideNotification persistence is deferred to a follow-up since the notification visibility logic involves additional surface area beyond this PR scope.
